toolkit/offchain/src/d_param.rs

```diff
@@ -1,13 +1,164 @@
+		Some((_, current_d_param)) if current_d_param == *d_parameter => {
+			log::info!("Current D-parameter value is equal to the one to be set.");
+			Ok(None)
+		},
+		Some((current_utxo, _)) => {
+			log::info!("Current D-parameter is different to the one to be set. Updating.");
+			Ok(Some(
+				update_d_param(&validator, &policy, d_parameter, &current_utxo, ctx, ogmios_client)
+					.await?,
+			))
+		},
+		None => {
+			log::info!("There is no D-parameter set. Inserting new one.");
+			Ok(Some(insert_d_param(&validator, &policy, d_parameter, ctx, ogmios_client).await?))
+		},
+	}
+}
+
+fn get_current_d_parameter(
+	validator_utxos: Vec<OgmiosUtxo>,
+) -> Result<Option<(OgmiosUtxo, DParameter)>, anyhow::Error> {
+	if let Some(utxo) = validator_utxos.first() {
+		let datum = utxo.datum.clone().ok_or_else(|| {
+			anyhow!("Invalid state: an UTXO at the validator script address does not have a datum")
+		})?;
+		let datum_plutus_data = PlutusData::from_bytes(datum.bytes).map_err(|e| {
+			anyhow!("Internal error: could not decode datum of D-parameter validator script: {}", e)
+		})?;
+		let current_d_param: DParameter =
+			DParamDatum::try_from(datum_plutus_data)
+				.map_err(|e| {
+					anyhow!("Internal error: could not decode datum of D-parameter validator script: {}", e)
+				})?
+				.into();
+		Ok(Some((utxo.clone(), current_d_param)))
+	} else {
+		Ok(None)
+	}
+}
+
+async fn insert_d_param<C>(
+	validator: &PlutusScript,
+	policy: &PlutusScript,
+	d_parameter: &DParameter,
+	ctx: TransactionContext,
+	client: &C,
+) -> anyhow::Result<McTxHash>
+where
+	C: Transactions,
+{
+	let zero_ex_units = ExUnits::new(&0u64.into(), &0u64.into());
+	let tx = mint_d_param_token_tx(validator, policy, d_parameter, &ctx, zero_ex_units)?;
+	let evaluate_response = client.evaluate_transaction(&tx.to_bytes()).await.map_err(|e| {
+		anyhow!(
+			"Evaluate insert D-parameter transaction request failed: {}, bytes: {}",
+			e,
+			hex::encode(tx.to_bytes())
+		)
+	})?;
+	let mint_witness_ex_units = get_mint_ex_units(evaluate_response)?;
+	let tx = mint_d_param_token_tx(validator, policy, d_parameter, &ctx, mint_witness_ex_units)?;
+	let signed_tx = ctx.sign(&tx).to_bytes();
+	let res = client.submit_transaction(&signed_tx).await.map_err(|e| {
+		anyhow!(
+			"Submit insert D-parameter transaction request failed: {}, bytes: {}",
+			e,
+			hex::encode(tx.to_bytes())
+		)
+	})?;
+	log::info!("Transaction submitted: {}", hex::encode(res.transaction.id));
+	// TODO: await for the transaction output to be visible in validator wallet
+	Ok(McTxHash(res.transaction.id))
+}
+
+fn get_mint_ex_units(response: Vec<OgmiosEvaluateTransactionResponse>) -> anyhow::Result<ExUnits> {
+	// For mint transaction we know there is a single item to evaluate, so matching is trivial.
+	let validator_and_budget = response
+		.first()
+		.ok_or_else(|| anyhow!("Internal error: cannot use evaluateTransaction response"))?;
+	Ok(convert_ex_units(&validator_and_budget.budget))
+}
```

Comment:
Can I please have some clarification to why the argument/variable names are different from `get_spend_ex_units`?
`response` vs `costs`, `validator_and_budget` vs `cost`

Reply:
I can make them equal. For me both are equally good/bad.

Reply:
I can make them equal. For me both are equally good/bad.

I'll merge it and fix in the PR I have on top of it.